maas/client/bones/helpers.py

```diff
@@ -207,16 +209,32 @@ class LoginNotSupported(LoginError):
             credentials = await authenticate(
                 url.geturl(), username, password, insecure=insecure)
 
-    # Circular import.
-    from ..bones.helpers import fetch_api_description
-    description = await fetch_api_description(url, credentials, insecure)
+    description = await fetch_api_description(url, insecure)
 
     # Return a new (unsaved) profile.
     return Profile(
         name=url.netloc, url=url.geturl(), credentials=credentials,
         description=description)
 
 
+async def authenticate_with_macaroon(url, insecure=False):
+    """Login via macaroons and generate and return new API keys."""
+    executor = futures.ThreadPoolExecutor(max_workers=1)
+
+    def get_token():
+        client = httpbakery.Client()
+        resp = client.request(
+            'POST', '{}/account/?op=create_authorisation_token'.format(url),
+            verify=not insecure)
+        if resp.status_code != 200:
+            raise LoginError('Login failed: {}'.format(resp.text))
```

Comment:
So, what happens if you don't specify a username and password when talking to a server that doesn't support macaroons?

I don't see anything catching LoginError, so I would assume that it simply fails. It would be better if it would fall back to prompting for username and password in that case.

Reply:
This would complicate the (already complex) logic for handing different combinations of cmdline options and interactive parameters quite a bit.
Also, prompting for user input in the login function seems a bit wrong (as user interaction is dealt with at a higher level.

Reply:
Personally, I would drop interactive username input (and only require password input when username is provided, to avoid showing it in the cmdline), this would simplify the logic

Reply:
This sucks that a thread needs to be used to use macaroons? How much work would it be to use aiohttp to handle this.

Reply:
It requires quite a rewrite of bakery, since everything would have to be async.

Reply:
I guess for now, but I think at some point we need to convert this to asyncio.